File: Practice/Dynamic_Programming/q33/q33.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int t[16], p[16], dp[16];
int n, max_v;

int main(void){
    cin >> n;
    for(int i = 0; i < n; i++){
        cin >> t[i] >> p[i];
    }
    for(int i = n-1; i >= 0; i--){
        int time = t[i] + i;
        if(time <= n){
            dp[i] = max(p[i]+dp[time], max_v);
            max_v = dp[i];
        }else{
            dp[i] = max_v;
        }
    }
    cout << max_v << "\n";
}

File: Practice/Dynamic_Programming/q33/q33.py
n = int(input())
t = []
p = []
dp = [0]*(n+1)
for i in range(n):
    ti, pi = map(int, input().split())
    t.append(ti)
    p.append(pi)
max_v = 0
for i in range(n-1, -1, -1):
    time = t[i] + i
    if time <= n:
        dp[i] = max(p[i]+dp[time], max_v)
        max_v = dp[i]
    else:
        dp[i] = max_v
print(max_v)